File: day1/part2.py
#!/usr/bin/python3

# simple quick and dirty python solution, for verification

import sys

digits = {  "1":1, "2":2, "3":3, "4":4, "5":5, "6":6, "7":7, "8":8, "9":9, 
        "one":1, "two":2, "three":3, "four":4, "five":5, "six":6, "seven":7, "eight":8, "nine":9 }

sum=0
for line in sys.stdin.readlines():
    firstpos=len(line)
    lastpos=-1
    for w in digits:
        pos=line.find(w)
        if pos!=-1:
            if pos<firstpos:
                firstpos=pos
                fdigit=digits[w]
        pos=line.rfind(w)
        if pos!=-1:
            if pos>lastpos:
                lastpos=pos
                ldigit=digits[w]
    sum+=fdigit*10+ldigit        
print("sum:",sum)

File: Day1/part2.c
// AoC day1, C, touch each input byte only once
#include <ctype.h>
#include <stdio.h>
#include <string.h>

char* digits[] = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
int wordpos[] = { 0, 0, 0, 0, 0, 0, 0, 0, 0 };

int match_in_word(int c) {
	int match=0;
	for(int cw=0; cw<9; cw++) {
		if (digits[cw][wordpos[cw]]==c) wordpos[cw]++;
		else {
			if(digits[cw][0]==c) 
				wordpos[cw]=1;
			else
				wordpos[cw]=0;
		}
		if (digits[cw][wordpos[cw]]==0) {
			wordpos[cw]=0;
			match=cw+1;
		}
	}
	return match;
}

int main(void) {
	int ldig, fdig = -1;
	int c, sum = 0;
	int wd;
	
	while ((c=fgetc(stdin))!=EOF) {
		if (wd=match_in_word(c)) {
			ldig=wd;
			if (fdig==-1) fdig=ldig;
		}
		if (isdigit(c)) {
			ldig=c-'0';
			if (fdig==-1) fdig=ldig;
		} else if (c=='\n') {
			sum += 10*fdig + ldig;
			fdig = -1;
			memset(wordpos, 0, 9*sizeof(int));
		}
	} 
	printf("sum=%d\n", sum);
}
